File: Logram/Models/LogDocument.swift
import Foundation
import Observation

/// Main model: holds all parsed lines, statistics, and filter state
@Observable
final class LogDocument {
    // MARK: - Data
    var allLines: [LogLine] = []
    var filteredIndices: [Int] = []
    var fileName: String = ""
    var filePath: String = ""
    var fileSize: Int64 = 0
    var parser: LogParser?

    // MARK: - Header info
    var ubVersion: String?
    var hostInfo: String?
    var startEpochCS: Int64 = -1
    var endEpochCS: Int64 = -1

    // MARK: - Statistics
    var perThreadCount: [Int] = Array(repeating: 0, count: LogParser.maxThreads)
    var perLevelCount: [Int] = Array(repeating: 0, count: LogLevel.allCases.count)
    var activeThreads: [Int] = [] // thread indices that have events

    // MARK: - Filters
    var enabledLevels: Set<LogLevel> = Set(LogLevel.allCases)
    var enabledThreads: Set<Int> = Set(0..<LogParser.maxThreads)
    var searchPattern: String = ""
    var searchRegex: Bool = false

    /// "Focus on call" — when set, applyFilters keeps only lines within this
    /// inclusive range. Other filters (level/thread/text) still apply on top.
    /// Cleared by clearFocus() or by loading a new file.
    var focusRange: ClosedRange<Int>? = nil
    /// Thread of the focused call frame (for status display).
    var focusThread: Int? = nil
    /// Saved enabledThreads at focus time, restored by clearFocus().
    private var savedEnabledThreads: Set<Int>? = nil

    // MARK: - Selection
    var selectedLineId: Int?

    // MARK: - Method timing
    struct MethodTiming: Identifiable {
        let id: Int       // line index of enter (+)
        let thread: Int
        let durationMS: Double
        let method: String
    }
    var methodTimings: [MethodTiming] = []

    // MARK: - Loading

    var isLoading = false
    var loadingProgress: Double = 0

    /// Per-chunk parse result for parallel aggregation
    private struct ChunkResult: Sendable {
        let chunkIndex: Int
        let lines: [LogLine]
        let threadCounts: [Int]  // size: maxThreads
        let levelCounts: [Int]   // size: LogLevel.allCases.count
    }

    func load(from url: URL) async {
        isLoading = true
        loadingProgress = 0
        fileName = url.lastPathComponent
        filePath = url.path

        do {
            let attrs = try FileManager.default.attributesOfItem(atPath: url.path)
            fileSize = attrs[.size] as? Int64 ?? 0
        } catch {}

        do {
            // Memory-map the file — no copy into RAM
            let data = try Data(contentsOf: url, options: .mappedIfSafe)
            let totalBytes = data.count

            // Find newline positions at byte level
            let lineRanges = data.withUnsafeBytes { (rawBuf: UnsafeRawBufferPointer) -> [Range<Int>] in
                guard let base = rawBuf.baseAddress?.assumingMemoryBound(to: UInt8.self) else { return [] }
                var ranges: [Range<Int>] = []
                ranges.reserveCapacity(totalBytes / 80) // rough estimate
                var start = 0
                for i in 0..<totalBytes {
                    if base[i] == 0x0A { // '\n'
                        ranges.append(start..<i)
                        start = i + 1
                    }
                }
                if start < totalBytes {
                    ranges.append(start..<totalBytes)
                }
                return ranges
            }

            let total = lineRanges.count

            // Initialize parser from header lines
            let headerCount = min(total, 10)
            var headerLines: [String] = []
            headerLines.reserveCapacity(headerCount)
            for i in 0..<headerCount {
                let range = lineRanges[i]
                let line = data.withUnsafeBytes { (rawBuf: UnsafeRawBufferPointer) -> String in
                    let ptr = rawBuf.baseAddress!.assumingMemoryBound(to: UInt8.self).advanced(by: range.lowerBound)
                    let buf = UnsafeBufferPointer(start: ptr, count: range.count)
                    return String(decoding: buf, as: UTF8.self)
                }
                headerLines.append(line)
            }
            let logParser = LogParser(headerLines: headerLines)
            self.parser = logParser
            self.ubVersion = logParser.ubVersion
            self.hostInfo = logParser.hostInfo
            self.startEpochCS = logParser.startEpochCS

            // Parallel parsing across CPU cores
            let cpuCount = max(ProcessInfo.processInfo.activeProcessorCount, 1)
            let chunkSize = max(total / cpuCount, 1)

            var chunkRanges: [(start: Int, end: Int)] = []
            for c in 0..<cpuCount {
                let start = c * chunkSize
                let end = c == cpuCount - 1 ? total : (c + 1) * chunkSize
                if start < total {
                    chunkRanges.append((start, end))
                }
            }

            loadingProgress = 0.1

            // Parse chunks in parallel
            let results: [ChunkResult] = await withTaskGroup(of: ChunkResult.self) { group in
                for (ci, chunkRange) in chunkRanges.enumerated() {
                    group.addTask {
                        var lines: [LogLine] = []
                        lines.reserveCapacity(chunkRange.end - chunkRange.start)
                        var threadCounts = Array(repeating: 0, count: LogParser.maxThreads)
                        var levelCounts = Array(repeating: 0, count: LogLevel.allCases.count)

                        for i in chunkRange.start..<chunkRange.end {
                            let byteRange = lineRanges[i]
                            // Strip trailing CR if present
                            var endIdx = byteRange.upperBound
                            if endIdx > byteRange.lowerBound {
                                let lastByte = data.withUnsafeBytes { buf in
                                    buf.load(fromByteOffset: endIdx - 1, as: UInt8.self)
                                }
                                if lastByte == 0x0D { endIdx -= 1 }
                            }
                            let lineStr = data.withUnsafeBytes { (rawBuf: UnsafeRawBufferPointer) -> String in
                                let ptr = rawBuf.baseAddress!.assumingMemoryBound(to: UInt8.self).advanced(by: byteRange.lowerBound)
                                let count = endIdx - byteRange.lowerBound
                                let buf = UnsafeBufferPointer(start: ptr, count: count)
                                return String(decoding: buf, as: UTF8.self)
                            }

                            let line = logParser.parseLine(lineStr, index: i)
                            lines.append(line)
                            if line.level != .unknown && line.thread >= 0 {
                                threadCounts[line.thread] += 1
                                levelCounts[line.level.rawValue] += 1
                            }
                        }
                        return ChunkResult(
                            chunkIndex: ci, lines: lines,
                            threadCounts: threadCounts, levelCounts: levelCounts
                        )
                    }
                }

                var collected: [ChunkResult] = []
                collected.reserveCapacity(chunkRanges.count)
                for await result in group {
                    collected.append(result)
                }
                return collected.sorted { $0.chunkIndex < $1.chunkIndex }
            }

            loadingProgress = 0.9

            // Merge results
            var parsed: [LogLine] = []
            parsed.reserveCapacity(total)
            var mergedThreadCounts = Array(repeating: 0, count: LogParser.maxThreads)
            var mergedLevelCounts = Array(repeating: 0, count: LogLevel.allCases.count)

            for result in results {
                parsed.append(contentsOf: result.lines)
                for j in 0..<LogParser.maxThreads {
                    mergedThreadCounts[j] += result.threadCounts[j]
                }
                for j in 0..<LogLevel.allCases.count {
                    mergedLevelCounts[j] += result.levelCounts[j]
                }
            }

            perThreadCount = mergedThreadCounts
            perLevelCount = mergedLevelCounts

            // Find end epochCS from last valid line
            for i in stride(from: parsed.count - 1, through: 0, by: -1) {
                if parsed[i].epochCS >= 0 {
                    endEpochCS = parsed[i].epochCS
                    break
                }
            }

            // Compute active threads
            activeThreads = perThreadCount.enumerated()
                .filter { $0.element > 0 }
                .map { $0.offset }

            allLines = parsed
            focusRange = nil
            focusThread = nil
            savedEnabledThreads = nil
            applyFilters()
            loadingProgress = 1.0
            isLoading = false

        } catch {
            isLoading = false
            print("Failed to load log: \(error)")
        }
    }

    // MARK: - Filtering

    func applyFilters() {
        let levels = enabledLevels
        let threads = enabledThreads
        let pattern = searchPattern
        let useRegex = searchRegex
        let range = focusRange

        var regex: NSRegularExpression?
        if !pattern.isEmpty && useRegex {
            regex = try? NSRegularExpression(pattern: pattern, options: .caseInsensitive)
        }

        var indices: [Int] = []
        indices.reserveCapacity(allLines.count)

        let lower = range?.lowerBound ?? 0
        let upper = range?.upperBound ?? (allLines.count - 1)

        for i in lower...max(lower, upper) {
            if i >= allLines.count { break }
            let line = allLines[i]
            // Level filter
            guard levels.contains(line.level) else { continue }
            // Thread filter
            if line.thread >= 0 {
                guard threads.contains(line.thread) else { continue }
            }
            // Text filter
            if !pattern.isEmpty {
                if let re = regex {
                    let nsr = NSRange(line.raw.startIndex..., in: line.raw)
                    if re.firstMatch(in: line.raw, range: nsr) == nil { continue }
                } else {
                    if !line.raw.localizedCaseInsensitiveContains(pattern) { continue }
                }
            }
            indices.append(i)
        }

        filteredIndices = indices
    }

    // MARK: - Focus on Call

    /// Walk up from `lineId` on its thread, find the nearest enclosing Enter
    /// (`+`), then locate its matching Leave (`-`). Set `focusRange` so that
    /// the table shows only that call frame. Returns false if the line isn't
    /// inside any open call frame on its thread.
    @discardableResult
    func focusOnCall(lineId: Int) -> Bool {
        guard lineId >= 0, lineId < allLines.count else { return false }
        let line = allLines[lineId]
        let thread = line.thread
        guard thread >= 0 else { return false }

        // Walk up the same thread tracking nested closed frames; whenever we
        // emerge to stack-empty and see an Enter, that's an enclosing frame
        // for our line. We don't stop at the first one — we keep going to
        // the very top of the call stack (outermost Enter still containing
        // the line). That gives the user "from process start to finish",
        // not just the innermost call.
        var stack = 0
        var topLevel = -1
        let startProbe = (line.level == .enter) ? lineId - 1 : lineId - 1
        if line.level == .enter { topLevel = lineId }
        var i = startProbe
        while i >= 0 {
            let prev = allLines[i]
            if prev.thread == thread {
                if prev.level == .leave {
                    stack += 1
                } else if prev.level == .enter {
                    if stack > 0 {
                        stack -= 1
                    } else {
                        // Enter that still contains our line — possibly the
                        // outermost so far. Keep climbing in case there's an
                        // even more outer frame above.
                        topLevel = i
                    }
                }
            }
            i -= 1
        }
        if topLevel < 0 { return false }
        let startIdx = topLevel

        // Find matching Leave; if there isn't one (truncated log), focus to EOF.
        let endIdx = findMatchingPair(for: startIdx) ?? (allLines.count - 1)
        focusRange = startIdx...endIdx
        focusThread = thread

        // Snapshot the user's thread filter so we can restore it on clearFocus,
        // then narrow to just the focused thread — that's how the user wants
        // it: the call frame visually contains only the originating thread.
        if savedEnabledThreads == nil {
            savedEnabledThreads = enabledThreads
        }
        enabledThreads = [thread]

        applyFilters()
        return true
    }

    func clearFocus() {
        guard focusRange != nil else { return }
        focusRange = nil
        focusThread = nil
        if let saved = savedEnabledThreads {
            enabledThreads = saved
            savedEnabledThreads = nil
        }
        applyFilters()
    }

    // MARK: - Search

    enum SearchDirection { case forward, backward }

    func findNext(_ pattern: String, direction: SearchDirection = .forward, from: Int? = nil) -> Int? {
        let indices = filteredIndices
        guard !indices.isEmpty, !pattern.isEmpty else { return nil }
        let re = try? NSRegularExpression(pattern: pattern, options: .caseInsensitive)
        let startIdx = from ?? (direction == .forward ? -1 : indices.count)

        if direction == .forward {
            for i in (startIdx + 1)..<indices.count {
                if matchLine(allLines[indices[i]].raw, pattern: pattern, regex: re) {
                    return i
                }
            }
        } else {
            let upper = min(startIdx, indices.count) - 1
            for i in stride(from: upper, through: 0, by: -1) {
                if matchLine(allLines[indices[i]].raw, pattern: pattern, regex: re) {
                    return i
                }
            }
        }
        return nil
    }

    private func matchLine(_ line: String, pattern: String, regex: NSRegularExpression?) -> Bool {
        if let re = regex {
            let range = NSRange(line.startIndex..., in: line)
            return re.firstMatch(in: line, range: range) != nil
        }
        return line.localizedCaseInsensitiveContains(pattern)
    }

    // MARK: - Method Timing

    func buildMethodTimings() {
        guard parser != nil else { return }
        var timings = [MethodTiming]()

        let lines = allLines
        let count = lines.count

        for i in 0..<count {
            guard lines[i].level == .enter else { continue }
            let th = lines[i].thread
            var depth = 0

            for j in (i + 1)..<count {
                guard lines[j].thread == th else { continue }
                if lines[j].level == .enter {
                    depth += 1
                } else if lines[j].level == .leave {
                    if depth == 0 {
                        let csStart = lines[i].epochCS
                        let csEnd = lines[j].epochCS
                        if csStart >= 0 && csEnd >= 0 {
                            let durationMS = Double(csEnd - csStart) * 10.0
                            if durationMS >= 10 {
                                let method = lines[i].message.trimmingCharacters(in: .whitespaces)
                                timings.append(MethodTiming(
                                    id: i, thread: th,
                                    durationMS: durationMS,
                                    method: method
                                ))
                            }
                        }
                        break
                    } else {
                        depth -= 1
                    }
                }
            }
        }

        timings.sort { $0.durationMS > $1.durationMS }
        methodTimings = timings

        // Write duration to Enter lines so the Duration column shows them
        for timing in timings {
            allLines[timing.id].durationUS = Int64(timing.durationMS * 1000)
        }
    }

    // MARK: - Jump to matching +/-

    /// Find the matching Enter for a Leave line, or matching Leave for an Enter line
    func findMatchingPair(for lineId: Int) -> Int? {
        guard lineId >= 0, lineId < allLines.count else { return nil }
        let line = allLines[lineId]
        let th = line.thread
        guard th >= 0 else { return nil }

        if line.level == .enter {
            var depth = 0
            for i in (lineId + 1)..<allLines.count {
                let l = allLines[i]
                guard l.thread == th else { continue }
                if l.level == .enter { depth += 1 }
                else if l.level == .leave {
                    if depth == 0 { return i }
                    depth -= 1
                }
            }
        } else if line.level == .leave {
            var depth = 0
            for i in stride(from: lineId - 1, through: 0, by: -1) {
                let l = allLines[i]
                guard l.thread == th else { continue }
                if l.level == .leave { depth += 1 }
                else if l.level == .enter {
                    if depth == 0 { return i }
                    depth -= 1
                }
            }
        }
        return nil
    }

    // MARK: - Computed stats

    var totalEvents: Int { allLines.count }

    var filteredCount: Int { filteredIndices.count }

    var httpRequests: Int {
        perLevelCount[LogLevel.http.rawValue]
    }

    var sqlQueries: Int {
        perLevelCount[LogLevel.sql.rawValue]
    }

    var errorCount: Int {
        perLevelCount[LogLevel.error.rawValue] +
        perLevelCount[LogLevel.exc.rawValue] +
        perLevelCount[LogLevel.excOs.rawValue] +
        perLevelCount[LogLevel.osErr.rawValue] +
        perLevelCount[LogLevel.fail.rawValue]
    }

    var duration: TimeInterval? {
        guard startEpochCS >= 0, endEpochCS >= 0 else { return nil }
        return Double(endEpochCS - startEpochCS) / 100.0
    }

    var durationFormatted: String {
        guard let d = duration else { return "—" }
        let h = Int(d) / 3600
        let m = (Int(d) % 3600) / 60
        let s = Int(d) % 60
        return String(format: "%d:%02d:%02d", h, m, s)
    }
}

File: LogramWin/src/core/LogDocument.cpp
#include "core/LogDocument.h"
#include <thread>
#include <algorithm>
#include <cstring>
#include <cmath>

std::string LogDocument::FileName() const {
    auto& p = file_.Path();
    auto pos = p.find_last_of(L"\\/");
    if (pos == std::wstring::npos) pos = 0; else pos++;
    std::string result;
    for (size_t i = pos; i < p.size(); ++i)
        result.push_back(static_cast<char>(p[i]));
    return result;
}

int LogDocument::ErrorCount() const {
    return perLevelCount_[static_cast<int>(LogLevel::Error)] +
           perLevelCount_[static_cast<int>(LogLevel::Exc)] +
           perLevelCount_[static_cast<int>(LogLevel::ExcOs)] +
           perLevelCount_[static_cast<int>(LogLevel::OsErr)] +
           perLevelCount_[static_cast<int>(LogLevel::Fail)];
}

double LogDocument::DurationSeconds() const {
    if (startEpochCS_ < 0 || endEpochCS_ < 0) return -1.0;
    return static_cast<double>(endEpochCS_ - startEpochCS_) / 100.0;
}

std::string LogDocument::DurationFormatted() const {
    double d = DurationSeconds();
    if (d < 0) return "\xe2\x80\x94"; // em-dash
    int total = static_cast<int>(d);
    int h = total / 3600;
    int m = (total % 3600) / 60;
    int s = total % 60;
    char buf[16];
    snprintf(buf, sizeof(buf), "%d:%02d:%02d", h, m, s);
    return buf;
}

int64_t LogDocument::GetDuration(uint32_t lineId) const {
    // Binary search in sorted durationLines_
    auto it = std::lower_bound(durationLines_.begin(), durationLines_.end(), lineId,
        [](const LogLineDuration& d, uint32_t id) { return d.lineId < id; });
    if (it != durationLines_.end() && it->lineId == lineId) return it->durationUS;
    return -1;
}

const LogLineHttp* LogDocument::GetHttp(uint32_t lineId) const {
    auto it = std::lower_bound(httpLines_.begin(), httpLines_.end(), lineId,
        [](const LogLineHttp& h, uint32_t id) { return h.lineId < id; });
    if (it != httpLines_.end() && it->lineId == lineId) return &(*it);
    return nullptr;
}

bool LogDocument::Load(const wchar_t* path, std::function<void(double)> onProgress) {
    MappedFile newFile;
    if (!newFile.Open(path)) return false;

    const uint8_t* base = newFile.Data();
    uint64_t totalBytes = newFile.Size();

    // Find newline positions
    std::vector<uint32_t> lineStarts;
    lineStarts.reserve(static_cast<size_t>(totalBytes / 80));
    lineStarts.push_back(0);
    for (uint64_t i = 0; i < totalBytes; ++i) {
        if (base[i] == 0x0A) {
            if (i + 1 < totalBytes)
                lineStarts.push_back(static_cast<uint32_t>(i + 1));
        }
    }

    size_t totalLines = lineStarts.size();

    // Compute line lengths (strip \r)
    struct LineRange { uint32_t offset; uint32_t length; };
    std::vector<LineRange> ranges(totalLines);
    for (size_t i = 0; i < totalLines; ++i) {
        uint32_t start = lineStarts[i];
        uint32_t end;
        if (i + 1 < totalLines) {
            end = lineStarts[i + 1] - 1; // before \n
        } else {
            end = static_cast<uint32_t>(totalBytes);
        }
        // Strip trailing \r
        if (end > start && base[end - 1] == 0x0D) --end;
        ranges[i] = {start, end - start};
    }

    // Initialize parser from header lines
    size_t headerCount = std::min(totalLines, size_t(10));
    std::vector<std::string_view> headerLines;
    headerLines.reserve(headerCount);
    for (size_t i = 0; i < headerCount; ++i) {
        headerLines.emplace_back(
            reinterpret_cast<const char*>(base + ranges[i].offset),
            ranges[i].length);
    }
    auto logParser = std::make_unique<LogParser>(headerLines);

    if (onProgress) onProgress(0.1);

    // Parallel parsing
    size_t cpuCount = std::max(1u, std::thread::hardware_concurrency());
    size_t chunkSize = std::max(totalLines / cpuCount, size_t(1));

    struct ChunkResult {
        std::array<int, kMaxThreads> threadCounts{};
        std::array<int, kLogLevelCount> levelCounts{};
        std::vector<LogLineHttp> httpLines;
        std::vector<LogLineDuration> durationLines;
    };

    std::vector<LogLineHot> parsed(totalLines);
    std::vector<ChunkResult> chunkResults(cpuCount);
    std::atomic<size_t> progressCounter{0};

    auto parseChunk = [&](size_t chunkIdx) {
        size_t start = chunkIdx * chunkSize;
        size_t end = (chunkIdx == cpuCount - 1) ? totalLines : (chunkIdx + 1) * chunkSize;
        if (start >= totalLines) return;
        end = std::min(end, totalLines);

        auto& cr = chunkResults[chunkIdx];

        for (size_t i = start; i < end; ++i) {
            auto& r = ranges[i];
            auto result = logParser->ParseLine(
                base + r.offset, r.length, r.offset, static_cast<uint32_t>(i));

            parsed[i] = result.hot;

            auto level = static_cast<LogLevel>(result.hot.level);
            int thread = result.hot.thread;

            if (level != LogLevel::Unknown && thread >= 0) {
                cr.threadCounts[thread]++;
                cr.levelCounts[static_cast<int>(level)]++;
            }

            if (result.durationUS >= 0) {
                cr.durationLines.push_back({static_cast<uint32_t>(i), result.durationUS});
            }

            if (level == LogLevel::Http && result.hot.messageOffset < r.length) {
                LogLineHttp http;
                http.lineId = static_cast<uint32_t>(i);
                LogParser::ParseHTTPFields(
                    base + r.offset, r.length, result.hot.messageOffset,
                    reinterpret_cast<const char*>(base + r.offset), http);
                if (!http.method.empty() || http.status >= 0) {
                    cr.httpLines.push_back(std::move(http));
                }
            }

            if ((i & 0x3FFF) == 0) { // every 16K lines
                progressCounter.store(i, std::memory_order_relaxed);
            }
        }
    };

    std::vector<std::jthread> threads;
    for (size_t c = 1; c < cpuCount; ++c) {
        threads.emplace_back([&, c] { parseChunk(c); });
    }
    parseChunk(0); // main thread does chunk 0
    threads.clear(); // join all

    if (onProgress) onProgress(0.9);

    // Merge results
    std::array<int, kMaxThreads> mergedThreadCounts{};
    std::array<int, kLogLevelCount> mergedLevelCounts{};
    std::vector<LogLineHttp> mergedHttp;
    std::vector<LogLineDuration> mergedDuration;

    for (auto& cr : chunkResults) {
        for (int j = 0; j < kMaxThreads; ++j) mergedThreadCounts[j] += cr.threadCounts[j];
        for (int j = 0; j < kLogLevelCount; ++j) mergedLevelCounts[j] += cr.levelCounts[j];
        mergedHttp.insert(mergedHttp.end(),
                          std::make_move_iterator(cr.httpLines.begin()),
                          std::make_move_iterator(cr.httpLines.end()));
        mergedDuration.insert(mergedDuration.end(),
                              cr.durationLines.begin(), cr.durationLines.end());
    }

    // Sort sparse arrays by lineId for binary search
    std::sort(mergedHttp.begin(), mergedHttp.end(),
              [](const LogLineHttp& a, const LogLineHttp& b) { return a.lineId < b.lineId; });
    std::sort(mergedDuration.begin(), mergedDuration.end(),
              [](const LogLineDuration& a, const LogLineDuration& b) { return a.lineId < b.lineId; });

    // Find end epochCS
    int64_t endCS = -1;
    for (size_t i = parsed.size(); i > 0; --i) {
        if (parsed[i - 1].epochCS >= 0) {
            endCS = parsed[i - 1].epochCS;
            break;
        }
    }

    // Active threads
    std::vector<int> active;
    for (int i = 0; i < kMaxThreads; ++i) {
        if (mergedThreadCounts[i] > 0) active.push_back(i);
    }

    // Commit
    file_ = std::move(newFile);
    parser_ = std::move(logParser);
    allLines_ = std::move(parsed);
    httpLines_ = std::move(mergedHttp);
    durationLines_ = std::move(mergedDuration);
    perThreadCount_ = mergedThreadCounts;
    perLevelCount_ = mergedLevelCounts;
    activeThreads_ = std::move(active);
    startEpochCS_ = parser_->GetStartEpochCS();
    endEpochCS_ = endCS;

    enabledLevelMask_ = ~uint64_t(0);
    enabledThreadMask_ = ~uint64_t(0);
    searchPattern_.clear();
    searchRegex_ = false;
    selectedLineId_ = -1;
    methodTimings_.clear();

    ApplyFilters();

    if (onProgress) onProgress(1.0);
    return true;
}

void LogDocument::SetEnabledLevelMask(uint64_t mask) { enabledLevelMask_ = mask; }
void LogDocument::SetEnabledThreadMask(uint64_t mask) { enabledThreadMask_ = mask; }
void LogDocument::SetSearchPattern(const std::string& pattern, bool isRegex) {
    searchPattern_ = pattern;
    searchRegex_ = isRegex;
}
void LogDocument::SetSelectedLineId(int id) { selectedLineId_ = id; }

void LogDocument::ApplyFilters() {
    uint64_t levelMask = enabledLevelMask_;
    uint64_t threadMask = enabledThreadMask_;
    const auto& pattern = searchPattern_;
    bool useRegex = searchRegex_;

    std::regex* re = nullptr;
    std::regex compiledRegex;
    if (!pattern.empty() && useRegex) {
        try {
            compiledRegex = std::regex(pattern, std::regex_constants::icase);
            re = &compiledRegex;
        } catch (...) {}
    }

    std::vector<uint32_t> indices;
    indices.reserve(allLines_.size());

    const uint8_t* base = file_.Data();

    for (size_t i = 0; i < allLines_.size(); ++i) {
        const auto& line = allLines_[i];
        auto level = static_cast<LogLevel>(line.level);

        // Level filter
        if (!(levelMask & (uint64_t(1) << static_cast<int>(level)))) continue;

        // Thread filter
        if (line.thread >= 0) {
            if (!(threadMask & (uint64_t(1) << line.thread))) continue;
        }

        // Text filter
        if (!pattern.empty()) {
            auto raw = GetRawLine(base, line);
            if (!MatchLine(raw, pattern, re)) continue;
        }

        indices.push_back(static_cast<uint32_t>(i));
    }

    filteredIndices_ = std::move(indices);
}

bool LogDocument::MatchLine(std::string_view raw, const std::string& pattern,
                            const std::regex* re) const {
    if (re) {
        return std::regex_search(raw.begin(), raw.end(), *re);
    }
    // Case-insensitive ASCII contains
    if (pattern.empty()) return true;
    auto it = std::search(raw.begin(), raw.end(), pattern.begin(), pattern.end(),
        [](char a, char b) {
            return (a >= 'A' && a <= 'Z' ? a + 32 : a) == (b >= 'A' && b <= 'Z' ? b + 32 : b);
        });
    return it != raw.end();
}

int LogDocument::FindMatchingPair(int lineId) const {
    if (lineId < 0 || lineId >= static_cast<int>(allLines_.size())) return -1;
    const auto& line = allLines_[lineId];
    auto level = static_cast<LogLevel>(line.level);
    int th = line.thread;
    if (th < 0) return -1;

    if (level == LogLevel::Enter) {
        int depth = 0;
        for (size_t i = lineId + 1; i < allLines_.size(); ++i) {
            const auto& l = allLines_[i];
            if (l.thread != th) continue;
            auto lv = static_cast<LogLevel>(l.level);
            if (lv == LogLevel::Enter) {
                depth++;
            } else if (lv == LogLevel::Leave) {
                if (depth == 0) return static_cast<int>(i);
                depth--;
            }
        }
    } else if (level == LogLevel::Leave) {
        int depth = 0;
        for (int i = lineId - 1; i >= 0; --i) {
            const auto& l = allLines_[i];
            if (l.thread != th) continue;
            auto lv = static_cast<LogLevel>(l.level);
            if (lv == LogLevel::Leave) {
                depth++;
            } else if (lv == LogLevel::Enter) {
                if (depth == 0) return i;
                depth--;
            }
        }
    }
    return -1;
}

void LogDocument::BuildMethodTimings() {
    methodTimings_.clear();
    if (!parser_) return;

    const uint8_t* base = file_.Data();
    size_t count = allLines_.size();

    // Single-pass O(n) with per-thread stacks.
    // Each stack entry is the line index of an Enter.
    std::array<std::vector<uint32_t>, kMaxThreads> stacks;

    for (size_t i = 0; i < count; ++i) {
        auto lv = static_cast<LogLevel>(allLines_[i].level);
        int th = allLines_[i].thread;
        if (th < 0 || th >= kMaxThreads) continue;

        if (lv == LogLevel::Enter) {
            stacks[th].push_back(static_cast<uint32_t>(i));
        } else if (lv == LogLevel::Leave && !stacks[th].empty()) {
            uint32_t enterIdx = stacks[th].back();
            stacks[th].pop_back();

            // Only top-level methods (stack now empty = outermost call)
            if (!stacks[th].empty()) continue;

            int64_t csStart = allLines_[enterIdx].epochCS;
            int64_t csEnd = allLines_[i].epochCS;
            if (csStart >= 0 && csEnd >= 0) {
                double durationMS = static_cast<double>(csEnd - csStart) * 10.0;
                if (durationMS >= 10.0) {
                    auto msg = GetMessage(base, allLines_[enterIdx]);
                    while (!msg.empty() && (msg.front() == ' ' || msg.front() == '\t'))
                        msg.remove_prefix(1);
                    while (!msg.empty() && (msg.back() == ' ' || msg.back() == '\t' ||
                                             msg.back() == '\r' || msg.back() == '\n'))
                        msg.remove_suffix(1);

                    methodTimings_.push_back({
                        enterIdx, th, durationMS, std::string(msg)
                    });
                }
            }
        }
    }

    std::sort(methodTimings_.begin(), methodTimings_.end(),
              [](const MethodTiming& a, const MethodTiming& b) {
                  return a.durationMS > b.durationMS;
              });
}

int LogDocument::FindNext(const std::string& pattern, SearchDirection dir, int from) const {
    const auto& indices = filteredIndices_;
    if (indices.empty() || pattern.empty()) return -1;

    std::regex* re = nullptr;
    std::regex compiledRegex;
    try {
        compiledRegex = std::regex(pattern, std::regex_constants::icase);
        re = &compiledRegex;
    } catch (...) {}

    const uint8_t* base = file_.Data();

    if (dir == SearchDirection::Forward) {
        int startIdx = from + 1;
        for (int i = startIdx; i < static_cast<int>(indices.size()); ++i) {
            auto raw = GetRawLine(base, allLines_[indices[i]]);
            if (MatchLine(raw, pattern, re)) return i;
        }
    } else {
        int upper = std::min(from, static_cast<int>(indices.size())) - 1;
        for (int i = upper; i >= 0; --i) {
            auto raw = GetRawLine(base, allLines_[indices[i]]);
            if (MatchLine(raw, pattern, re)) return i;
        }
    }
    return -1;
}
